src/Components/Login.js

```diff
@@ -0,0 +1,69 @@
+import React, { Component } from 'react';
+
+const uri = process.env.REACT_APP_API_ENDPOINT || 'http://localhost:8080'
+
+class Login extends Component {
+
+  state = {
+    username: '',
+    player: null
+  }
+
+  componentDidMount() {
+    this.loginStatus();
+  }
+
+  loginStatus() {
+    return fetch(`${uri}/login`, {
+      method: 'GET',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      credentials: 'include'
+    }).then(res => {
+      if(res.status < 400) return res.json()
+      else return null
+    }).then(user => {
+      this.setState({...this.state, player: user})
+      if(this.props.onLogin) {
+        this.props.onLogin(this.state.player);
+      }
+    })
+  }
+
+  loginAs(username) {
+    return fetch(`${uri}/login?user=${username}`, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      credentials: 'include'
+    }).then(res => this.loginStatus())
+  }
+
+  render() {
+    if(this.state.player == null) {
```

Comment:
Instead of having Login() have the player state, I think this should be a prop. So the flow is:

Main.js creates this component and passes it the player state, which will initially be null. It also passes it a `onLogin()` prop, which you already have. When Main.js.onLogin() is called, it updates it's state to have the player. This forces Main.js to re-render, and it would then pass in the player to onLogin().

This will simplify the state, and make Login.js really dumb (in a good way).

Reply:
That's the dream.